### PR TITLE
Interpolation with precomputed coefficients

### PR DESCRIPTION
**Pull request by navjotk**
This PR:
- Adds (brings back) a `FixedDimension` class, used when the size of a dimension is known at the time of creation
- Adds a `PrecomputedSparseFunction` class that can do interpolation of sparse points using precomputed coefficients - this can be used to implement arbitrary interpolation schemes. 
- Adds a test for the new `PrecomputedSparseFunction` class by precomputing coefficients using a bilinear scheme

~This branch currently does not pass tests due to a DSE bug where symbols used as indices for `Indexed` types are not found by the visitors. @FabioLuporini could you look into this please. An alternate branch `interpolation2` has a few hacks in place to work around the bug and passes this test.~

### CODE REVIEW

File: devito/dimension.py [hunk 1]
```diff
@@ -155,6 +156,21 @@ class TimeDimension(Dimension):
     """
 
 
+class FixedDimension(Dimension):
+    is_Default = True
```
Comment (FabioLuporini) [inline devito/dimension.py]: broken

File: devito/dimension.py [hunk 2]
```diff
@@ -155,6 +156,21 @@ class TimeDimension(Dimension):
...
+
+    def __new__(cls, name, **kwargs):
+        newobj = sympy.Symbol.__new__(cls, name)
+        newobj._reverse = kwargs.get('reverse', False)
```
Comment (FabioLuporini) [inline devito/dimension.py]: things changed in Devito

File: devito/dimension.py [hunk 3]
```diff
@@ -155,6 +156,21 @@ class TimeDimension(Dimension):
...
+        newobj._spacing = kwargs.get('spacing', Scalar(name='h_%s' % name))
+        newobj.default_value = kwargs.get('default_value', None)
```
Comment (FabioLuporini) [inline devito/dimension.py]: if it's fixed, "default" is misleading

File: devito/function.py [hunk 4]
```diff
@@ -1089,3 +1089,178 @@ def __indices_setup__(cls, **kwargs):
     @classmethod
     def __shape_setup__(cls, **kwargs):
         return kwargs.get('shape', (kwargs.get('nt'), kwargs.get('npoint'),))
+
+
+class PrecomputedSparseFunction(TensorFunction):
```
Comment (FabioLuporini) [inline devito/function.py]: I think the two ``SparseFunction`` classes should share some code

File: tests/test_interpolation.py [hunk 5]
```diff
@@ -1,9 +1,12 @@
 import numpy as np
 import pytest
 from conftest import skipif_yask
+from math import sin
+from math import floor
```
Comment (FabioLuporini) [inline tests/test_interpolation.py]: from math import sin, floor

**Comment by FabioLuporini** [comment]
> @FabioLuporini could you look into this please.

Sure, welcome to my stack of TODOs :-)

File: devito/dimension.py (hunk 3 above)
Comment (navjotk) [reply devito/dimension.py]: Default is a better name than fixed.

Comment (FabioLuporini) [reply devito/dimension.py]: makes me think the value can actually change. So it wouldn't be "fixed"

Comment (mloubout) [reply devito/dimension.py]: Why not just value?

Comment (navjotk) [reply devito/dimension.py]: We can always provide a runtime value to `Operator.apply` and the new value would be used, so it isn't fixed. Which is also why it makes sense to call the constructor argument `default_value` since it's the value to be used when one is not provided at runtime.

File: devito/dimension.py [hunk 6]
```diff
@@ -369,13 +384,3 @@ def origin(self):
 
     def _hashable_content(self):
         return Symbol._hashable_content(self) + (self.origin,)
-
-
-def dimensions(names):
```
Comment (FabioLuporini) [inline devito/dimension.py]: why?

Comment (navjotk) [reply devito/dimension.py]: Mistake

**Comment by navjotk** [comment]
A normal Dimension is associated with a `Function` or `SparseFunction` object and uses its shape to determine its default sizes. In this case, we have a dimension that is not associated with any `Function`/`SparseFunction` object, but its size is known at the time it is being created so it can be provided as a default value. 

I don't mind making it fixed. I think this added flexibility (of changing the size of the dimension - and hence the interpolator - at runtime) is good.

**Comment by ggorman** [comment]
What remains to land this?

**Comment by FabioLuporini** [comment]
> What remains to land this?

none of my comments have been addressed so far (feel free to have a look) and there are two bugs that might be fixed using the machinery in #517 ; a hacky fix was proposed by Navjot in a separate branch (``interpolation2``)

File: tests/test_interpolation.py [hunk 7]
```diff
@@ -50,6 +52,48 @@ def custom_points(grid, ranges, npoints, name='points'):
...
+    for i, point in enumerate(points):
+        for d in range(grid.dim):
+            coefficients[i, d, 0] = ((gridpoints[i][d] + 1)*grid.spacing[d] -
+                                     point[d])/grid.spacing[d]
+            coefficients[i, d, 1] = (point[d]-gridpoints[i][d]*grid.spacing[d])\
+                / grid.spacing[d]
+    return gridpoints, coefficients
+
+
+def test_precomputed_interpolation():
```
Comment (FabioLuporini) [inline tests/test_interpolation.py]: docstring

File: devito/function.py [hunk 8]
```diff
@@ -815,6 +781,88 @@ def __indices_setup__(cls, **kwargs):
...
+
+        :param alias: (Optional) name under which to store values.
+        """
+        key = alias or self
+        args = super(AbstractSparseFunction, self).argument_defaults(alias=alias)
+        for child in self.children:
+            args.update(child.argument_defaults(alias=getattr(key, child.name).name))
+        return args
+
+    def argument_values(self, alias=None, **kwargs):
```
Comment (mloubout) [inline devito/function.py]: Why is this needed back?

File: devito/function.py [hunk 9]
```diff
@@ -737,39 +737,11 @@ def _arg_values(self, alias=None, **kwargs):
...
-
-    .. note::
-
-        The parameters must always be given as keyword arguments, since
-        SymPy uses `*args` to (re-)create the dimension arguments of the
-        symbolic function.
-    """
-
-    is_SparseFunction = True
+class AbstractSparseFunction(TensorFunction):
```
Comment (mloubout) [inline devito/function.py]: docstring?

File: devito/function.py [hunk 10]
```diff
@@ -1074,3 +1122,107 @@ def __indices_setup__(cls, **kwargs):
...
+            assert(gridpoints_data is not None)
+            gridpoints.data[:] = gridpoints_data[:]
+            self.gridpoints = gridpoints
+
+            coefficients = Function(name="%s_coefficients" % self.name,
+                                    dimensions=(self.indices[-1], Dimension(name='d'),
+                                                Dimension(name='i')),
+                                    shape=(self.npoint, self.grid.dim, self.r),
+                                    space_order=0)
+            coefficients_data = kwargs.get('coefficients', None)
```
Comment (mloubout) [inline devito/function.py]: So for gridpoints/coefficients, wouldn't it be more consistent to keep `coordinate` and have an `interpolator` as input that computes gridpoints and coefficients from the coordinates.
```
sf = PrecomputedSparseFunction(name='s', grid=grid, r=r, npoint=len(points),
                                                   coordinates=coordinate, inteprolator=precompute_linear_interpolation)
```

File: devito/function.py [hunk 11]
```diff
@@ -1074,3 +1122,107 @@ def __indices_setup__(cls, **kwargs):
...
+            t = self.grid.stepping_dim
+            expr = expr.subs(t, u_t).subs(time, u_t)
+
+        coefficients = self.coefficients.indexed
+        gridpoints = self.gridpoints.indexed
+        p, _, _ = self.coefficients.indices
+        dim_subs = []
+        coeffs = []
+        for i, d in enumerate(self.grid.dimensions):
+            rd = DefaultDimension(name="r%s" % d.name, default_value=self.r)
```
Comment (mloubout) [inline devito/function.py]: Dow we really need a DInmension for that. Unless go to like 100 points in each direction, can just write down the r equations.

File: devito/function.py [hunk 12]
```diff
@@ -1074,3 +1122,107 @@ def __indices_setup__(cls, **kwargs):
...
+            expr = expr.subs(self.indices[0], p_t)
+
+        gridpoints = self.gridpoints.indexed
+        coefficients = self.coefficients.indexed
+
+        p, _ = self.gridpoints.indices
+        dim_subs = []
+        coeffs = []
+        for i, d in enumerate(self.grid.dimensions):
+            rd = DefaultDimension(name="r%s" % d.name, default_value=self.r)
```
Comment (mloubout) [inline devito/function.py]: same

File: tests/test_interpolation.py [hunk 13]
```diff
@@ -50,6 +52,48 @@ def custom_points(grid, ranges, npoints, name='points'):
     return points
 
 
+def precompute_linear_interpolation(points, grid, origin):
```
Comment (mloubout) [inline tests/test_interpolation.py]: docstring?

File: tests/test_interpolation.py [hunk 14]
```diff
@@ -50,6 +52,48 @@ def custom_points(grid, ranges, npoints, name='points'):
...
+
+    gridpoints, coefficients = precompute_linear_interpolation(points, grid, origin)
+
+    sf = PrecomputedSparseFunction(name='s', grid=grid, r=r, npoint=len(points),
+                                   gridpoints=gridpoints, coefficients=coefficients)
+    eqn = sf.interpolate(m)
+    op = Operator(eqn)
+    op()
+    expected_values = [sin(point[0]) + sin(point[1]) for point in points]
+    assert(all(np.isclose(sf.data, expected_values, rtol=1e-6)))
```
Comment (mloubout) [inline tests/test_interpolation.py]: adjoint test for this?

File: devito/function.py [hunk 15]
```diff
@@ -815,6 +781,66 @@ def __indices_setup__(cls, **kwargs):
...
+    def _arg_defaults(self, alias=None):
+        """
+        Returns a map of default argument values defined by this symbol.
+
+        :param alias: (Optional) name under which to store values.
+        """
+        key = alias or self
+        args = super(AbstractSparseFunction, self)._arg_defaults(alias=alias)
+        for child in self.child_functions:
+            args.update(child._arg_defaults(alias=getattr(key, child.name).name))
```
Comment (mloubout) [inline devito/function.py]: `alias` should be `gettattr` not just the name.

File: devito/dimension.py [hunk 16]
```diff
@@ -226,6 +227,30 @@ class TimeDimension(Dimension):
...
+    is_Default = True
+
+    def __new__(cls, name, **kwargs):
+        newobj = sympy.Symbol.__new__(cls, name)
+        newobj._spacing = kwargs.get('spacing', Scalar(name='h_%s' % name))
+        newobj.default_value = kwargs.get('default_value', None)
+        return newobj
+
+    def _arg_defaults(self, start=None, size=None, alias=None):
+        """
```
Comment (FabioLuporini) [inline devito/dimension.py]: drop docstring, as if identical to the superclass', then it should just be inherited. I'm now trying to follow this approach to avoid proliferation of redundant docstrings

File: devito/dimension.py [hunk 17]
```diff
@@ -472,7 +497,6 @@ def _hashable_content(self):
 def dimensions(names):
     """
     Shortcut for: ::
-
```
Comment (FabioLuporini) [inline devito/dimension.py]: why? It might screw sphinx

File: devito/function.py [hunk 18]
```diff
@@ -785,21 +757,15 @@ def __init__(self, *args, **kwargs):
...
-            self.coordinates = coordinates
-
             # Halo region
             self._halo = tuple((0, 0) for i in range(self.ndim))
 
             # Padding region
             self._padding = tuple((0, 0) for i in range(self.ndim))
 
+            # Symbols that are encapsulated within this symbol (e.g. coordinates)
+            self.child_functions = []
```
Comment (FabioLuporini) [inline devito/function.py]: private

File: devito/function.py (hunk 10 above)
Comment (navjotk) [reply devito/function.py]: You also need grid spacing and origin to map from coordinates to gridpoints. Can't stay consistent if we have to add those values too.

File: devito/function.py (hunk 11 above)
Comment (navjotk) [reply devito/function.py]: `Dimension`s are the only way to structure loops in devito. With a loop, the generated code is independent of the interpolator used. 
This is a WIP anyway. I don't think one way is better than the other. The best would be to pass the size information to the compiler and let the compiler decide whether it wants to unroll the loop.

File: devito/function.py (hunk 18 above)
Comment (navjotk) [reply devito/function.py]: Fixed

File: devito/dimension.py (hunk 17 above)
Comment (navjotk) [reply devito/dimension.py]: Fixed

File: devito/dimension.py (hunk 16 above)
Comment (navjotk) [reply devito/dimension.py]: Fixed

File: devito/function.py (hunk 15 above)
Comment (navjotk) [reply devito/function.py]: Fixed

File: tests/test_interpolation.py (hunk 13 above)
Comment (navjotk) [reply tests/test_interpolation.py]: Done

File: tests/test_interpolation.py (hunk 7 above)
Comment (navjotk) [reply tests/test_interpolation.py]: Done

File: tests/test_interpolation.py (hunk 14 above)
Comment (navjotk) [reply tests/test_interpolation.py]: WIP

File: devito/function.py (hunk 10 above)
Comment (mloubout) [reply devito/function.py]: Why make this one the only Function that does not take a grid in?

Comment (navjotk) [reply devito/function.py]: It [does](https://github.com/opesci/devito/pull/515/files#diff-f3329261db6f5e1efaecf750bcfc52e9R95)?

Comment (mloubout) [reply devito/function.py]: If it follow every other function, you have a Grid, so you have spacing and origin

Comment (navjotk) [reply devito/function.py]: The grid only has *symbols* for spacing and origin which are replaced by values only at apply time. To do any precomputation, I need those *values* before apply time. Not possible with the grid object.

Comment (mloubout) [reply devito/function.py]: > I need those values before apply time.

You shouldn't , the `subs` is supposed to take care of it. Hard codingh values that may have nothing to do with the one that will be used at runtime is super risky.

Comment (mloubout) [reply devito/function.py]: And Grid always have spacing/origin values

Comment (navjotk) [reply devito/function.py]: Combining 
> You also need grid spacing and origin to map from coordinates to gridpoints.
with the fact that gridpoints need to be **pre**computed means I need spacing and origin before apply time. I don't know how else to explain this.
Hence, it's best to keep this mess outside of our classes and in user space - which is where this PR keeps it.

Comment (mloubout) [reply devito/function.py]: I find this really hacky and error prone.

Comment (navjotk) [reply devito/function.py]: This was raised in a meeting and implemented after agreement with you, @ggorman and @FabioLuporini .

Comment (navjotk) [reply devito/function.py]: Forcing the user to do the precomputation is better than silently using the origin and spacing values used to intialise the grid - since these might be changed at apply time, following your own logic set out in:
>Hard codingh values that may have nothing to do with the one that will be used at runtime is super risky.

File: devito/dimension.py [hunk 19]
```diff
@@ -226,6 +227,21 @@ class TimeDimension(Dimension):
     """
 
 
+class DefaultDimension(Dimension):
+    is_Default = True
+
+    def __new__(cls, name, **kwargs):
+        newobj = sympy.Symbol.__new__(cls, name)
+        newobj._spacing = kwargs.get('spacing', Scalar(name='h_%s' % name))
+        newobj.default_value = kwargs.get('default_value', None)
```
Comment (FabioLuporini) [inline devito/dimension.py]: should this be private?

File: devito/function.py [hunk 20]
```diff
@@ -815,6 +786,68 @@ def __indices_setup__(cls, **kwargs):
...
+    def _arg_defaults(self, alias=None):
+        """
+        Returns a map of default argument values defined by this symbol.
+
+        :param alias: (Optional) name under which to store values.
+        """
+        key = alias or self
+        args = super(AbstractSparseFunction, self)._arg_defaults(alias=alias)
+        for child_name in self._child_functions:
+            child = getattr(self, child_name)
```
Comment (FabioLuporini) [inline devito/function.py]: why don't you just put the actual functions in ``_child_functions`` rather than their names?

Comment (navjotk) [reply devito/function.py]: If I have a receiver symbol called `rec`, `self.coordinates.name` on it would give you `rec_coordinates`. If it is replaced by a symbol called `rec2`, the only way of accessing the coordinates `Function` on it is by doing `getattr(rec2, 'coordinates')` since (assuming `child = self.coordinates`) `getattr(rec2, child.name)` won't be found (the new symbol's coordinates function is automatically named `rec2_coordinates`.

Comment (navjotk) [reply devito/function.py]: Which also means this should probably be `child = getattr(key, child_name)`. Changing that.

Comment (navjotk) [reply devito/function.py]: This line: `args.update(child._arg_defaults(alias=getattr(key, child_name)))` will not work when storing the actual functions, only with the name, since the name is the only consistent way of accessing children across different `SparseFunction` instances.

File: devito/function.py (hunk 10 above)
Comment (mloubout) [reply devito/function.py]: Makes sense. Can we add some warnings at setup/runtime saying the setup is only valid for a specific grid?

This all PR is gonna be really handy for sparse-fd btw, so not trying to stop it by anyway, just making sure we don't making too different than the rest of SparseFunctions

File: devito/function.py (hunk 20 above)
Comment (mloubout) [reply devito/function.py]: This is really odd, `child._arg_defaults` should pick up the name itself and all `arg_defaults` take an object as alias, not a name.

Comment (FabioLuporini) [reply devito/function.py]: sorry, I didn't see `_child_function` was a class attribute. Now it does make sense

File: devito/function.py [hunk 21]
```diff
@@ -815,6 +786,68 @@ def __indices_setup__(cls, **kwargs):
...
+        if not self._cached():
+            super(SparseFunction, self).__init__(*args, **kwargs)
+
+            # Set up coordinates of sparse points
+            coordinates = Function(name='%s_coords' % self.name,
+                                   dimensions=(self.indices[-1], Dimension(name='d')),
+                                   shape=(self.npoint, self.grid.dim), space_order=0)
+            coordinate_data = kwargs.get('coordinates')
+            if coordinate_data is not None:
+                coordinates.data[:] = coordinate_data[:]
```
Comment (FabioLuporini) [inline devito/function.py]: random question unrelated to this PR: do we really want to copy data, or would we rather want a ``SparseFunction`` to take ownership of what it's been given?

File: devito/function.py [hunk 22]
```diff
@@ -1074,3 +1090,108 @@ def __indices_setup__(cls, **kwargs):
     @classmethod
     def __shape_setup__(cls, **kwargs):
         return kwargs.get('shape', (kwargs.get('nt'), kwargs.get('npoint'),))
+
+
+class PrecomputedSparseFunction(AbstractSparseFunction):
```
Comment (FabioLuporini) [inline devito/function.py]: we need a good docstring here

File: devito/function.py [hunk 23]
```diff
@@ -1074,3 +1090,108 @@ def __indices_setup__(cls, **kwargs):
...
+    is_PrecomputedSparseFunction = True
```
Comment (FabioLuporini) [inline devito/function.py]: we need ``is_PrecomputedSparseFunction = False`` in ``class Basic``, within ``types.py``

File: devito/function.py (hunk 10 above)
Comment (navjotk) [reply devito/function.py]: Sure.

File: devito/function.py (hunk 21 above)
Comment (navjotk) [reply devito/function.py]: I don't follow what you mean. Care to elaborate?

File: devito/function.py (hunk 20 above)
Comment (navjotk) [reply devito/function.py]: @mloubout there is no difference between what this line is doing and what is being done [here](https://github.com/opesci/devito/blob/master/devito/function.py#L1023), except here, it works for any arbitrarily named child attributes, not just `self.coordinates`. So, for a `PrecomputedSparseFunction`, this will execute:
`args.update(self.gridpoints._arg_defaults(alias=key.gridpoints)`. I believe I am reproducing the same behaviour as master, but in a more general way.

File: devito/dimension.py (hunk 19 above)
Comment (navjotk) [reply devito/dimension.py]: Sure

File: devito/function.py (hunk 22 above)
Comment (navjotk) [reply devito/function.py]: Done

File: devito/function.py (hunk 23 above)
Comment (navjotk) [reply devito/function.py]: Done
